### ISSUE TITLE
Code incompatible with bazel layering_check

### ISSUE THREAD

**Issue by iphydf** [protocolbuffers/protobuf#9406]
**What version of protobuf and what language are you using?**
Version: v3.19.3
Language: C++

**What operating system (Linux, Windows, ...) and version?**
Linux

**What runtime / compiler are you using (e.g., python version or gcc version)**
clang-14

**What did you do?**
1. Build a bazel project using protobuf with `--features=layering_check`.
2. See error

**What did you expect to see**
```
INFO: Build completed successfully
```

**What did you see instead?**
```
external/com_google_protobuf/src/google/protobuf/compiler/main.cc:31:10: error: module @com_google_protobuf//:protoc does not depend on a module exporting 'google/protobuf/compiler/cpp/cpp_generator.h'
#include <google/protobuf/compiler/cpp/cpp_generator.h>
         ^
external/com_google_protobuf/src/google/protobuf/compiler/main.cc:32:10: error: module @com_google_protobuf//:protoc does not depend on a module exporting 'google/protobuf/compiler/java/java_generator.h'
#include <google/protobuf/compiler/java/java_generator.h>
         ^
external/com_google_protobuf/src/google/protobuf/compiler/main.cc:33:10: error: module @com_google_protobuf//:protoc does not depend on a module exporting 'google/protobuf/compiler/java/java_kotlin_generator.h'
#include <google/protobuf/compiler/java/java_kotlin_generator.h>
         ^
external/com_google_protobuf/src/google/protobuf/compiler/main.cc:34:10: error: module @com_google_protobuf//:protoc does not depend on a module exporting 'google/protobuf/compiler/js/js_generator.h'
#include <google/protobuf/compiler/js/js_generator.h>
         ^
external/com_google_protobuf/src/google/protobuf/compiler/main.cc:35:10: error: module @com_google_protobuf//:protoc does not depend on a module exporting 'google/protobuf/compiler/command_line_interface.h'
#include <google/protobuf/compiler/command_line_interface.h>
         ^
external/com_google_protobuf/src/google/protobuf/compiler/main.cc:36:10: error: module @com_google_protobuf//:protoc does not depend on a module exporting 'google/protobuf/compiler/python/python_generator.h'
#include <google/protobuf/compiler/python/python_generator.h>
         ^
external/com_google_protobuf/src/google/protobuf/compiler/main.cc:37:10: error: module @com_google_protobuf//:protoc does not depend on a module exporting 'google/protobuf/compiler/csharp/csharp_generator.h'
#include <google/protobuf/compiler/csharp/csharp_generator.h>
         ^
external/com_google_protobuf/src/google/protobuf/compiler/main.cc:38:10: error: module @com_google_protobuf//:protoc does not depend on a module exporting 'google/protobuf/compiler/objectivec/objectivec_generator.h'
#include <google/protobuf/compiler/objectivec/objectivec_generator.h>
         ^
external/com_google_protobuf/src/google/protobuf/compiler/main.cc:39:10: error: module @com_google_protobuf//:protoc does not depend on a module exporting 'google/protobuf/compiler/php/php_generator.h'
#include <google/protobuf/compiler/php/php_generator.h>
         ^
external/com_google_protobuf/src/google/protobuf/compiler/main.cc:40:10: error: module @com_google_protobuf//:protoc does not depend on a module exporting 'google/protobuf/compiler/ruby/ruby_generator.h'
#include <google/protobuf/compiler/ruby/ruby_generator.h>
         ^
external/com_google_protobuf/src/google/protobuf/compiler/main.cc:42:10: error: module @com_google_protobuf//:protoc does not depend on a module exporting 'google/protobuf/port_def.inc'
#include <google/protobuf/port_def.inc>
         ^
11 errors generated.
```

**Anything else we should know about your project / environment**
My project is here: https://github.com/TokTok/toktok-stack

**Comment by deannagarcia** [comment]
I can't seem to reproduce this error, can you give me an exact command to reproduce?

**Comment by deannagarcia** [comment]
Closing this due to no response, if you're still seeing this error please open it back up.

**Comment by hzeller** [comment]
It is still an issue.

### To reproduce

Make sure compiler is `clang`. Only with clang, the `--features=layering_check` works in bazel (https://bazel.build/reference/be/c-cpp).

Here is a minimal project to reproduce. In an empty directory, create a WORKSPACE file and a .bazelrc to make sure we use c++17

```bash
cat <<EOF > WORKSPACE
workspace(name = "proto_testing")

load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

http_archive(
    name = "com_google_protobuf",
    sha256 = "8ff511a64fc46ee792d3fe49a5a1bcad6f7dc50dfbba5a28b0e5b979c17f9871",
    strip_prefix = "protobuf-25.2",
    urls = [
        "https://github.com/protocolbuffers/protobuf/releases/download/v25.2/protobuf-25.2.tar.gz",
    ],
)
load("@com_google_protobuf//:protobuf_deps.bzl", "protobuf_deps")
protobuf_deps()

http_archive(
    name = "rules_proto",
    sha256 = "dc3fb206a2cb3441b485eb1e423165b231235a1ea9b031b4433cf7bc1fa460dd",
    strip_prefix = "rules_proto-5.3.0-21.7",
    urls = [
        "https://github.com/bazelbuild/rules_proto/archive/refs/tags/5.3.0-21.7.tar.gz",
    ],
)

load("@rules_proto//proto:repositories.bzl", "rules_proto_dependencies")
rules_proto_dependencies()
EOF

cat <<EOF > .bazelrc
build --cxxopt=-xc++ --host_cxxopt=-xc++ --cxxopt=-std=c++17 --host_cxxopt=-std=c++17 --client_env=BAZEL_CXXOPTS=-std=c++17
EOF
```

Now, let's run it. Here, using bazel version 6.4.0 and clang++. Make sure to set the CXX variable to `clang++`, otherwise bazel will just use gcc and _not_ have the layering check feature.

```
bazel --version
bazel 6.4.0

$CXX --version
clang version 13.0.1
Target: x86_64-unknown-linux-gnu
```

Now run bazel with the layering check enabled:
```
bazel build --features=layering_check @com_google_protobuf//:any_proto --keep_going
```

Output is many hundrets of layering check messages, and of course the build fails.
```
external/com_google_protobuf/src/google/protobuf/descriptor.cc:34:10: error: module @com_google_protobuf//src/google/protobuf:protobuf_nowkt does not depend on a module exporting 'absl/functional/function_ref.h'
external/com_google_protobuf/src/google/protobuf/descriptor.cc:38:10: error: module @com_google_protobuf//src/google/protobuf:protobuf_nowkt does not depend on a module exporting 'absl/memory/memory.h'
external/com_google_protobuf/src/google/protobuf/descriptor.cc:39:10: error: module @com_google_protobuf//src/google/protobuf:protobuf_nowkt does not depend on a module exporting 'absl/status/statusor.h'
external/com_google_protobuf/src/google/protobuf/descriptor.cc:40:10: error: module @com_google_protobuf//src/google/protobuf:protobuf_nowkt does not depend on a module exporting 'absl/strings/ascii.h'
external/com_google_protobuf/src/google/protobuf/descriptor.cc:41:10: error: module @com_google_protobuf//src/google/protobuf:protobuf_nowkt does not depend on a module exporting 'absl/strings/escaping.h'
external/com_google_protobuf/src/google/protobuf/descriptor.cc:42:10: error: module @com_google_protobuf//src/google/protobuf:protobuf_nowkt does not depend on a module exporting 'absl/strings/match.h'
external/com_google_protobuf/src/google/protobuf/descriptor.cc:43:10: error: module @com_google_protobuf//src/google/protobuf:protobuf_nowkt does not depend on a module exporting 'absl/strings/str_cat.h'
external/com_google_protobuf/src/google/protobuf/descriptor.cc:44:10: error: module @com_google_protobuf//src/google/protobuf:protobuf_nowkt does not depend on a module exporting 'absl/strings/str_format.h'
external/com_google_protobuf/src/google/protobuf/descriptor.cc:45:10: error: module @com_google_protobuf//src/google/protobuf:protobuf_nowkt does not depend on a module exporting 'absl/strings/str_join.h'
external/com_google_protobuf/src/google/protobuf/descriptor.cc:46:10: error: module @com_google_protobuf//src/google/protobuf:protobuf_nowkt does not depend on a module exporting 'absl/strings/str_split.h'
external/com_google_protobuf/src/google/protobuf/descriptor.cc:47:10: error: module @com_google_protobuf//src/google/protobuf:protobuf_nowkt does not depend on a module exporting 'absl/strings/string_view.h'
external/com_google_protobuf/src/google/protobuf/descriptor.cc:48:10: error: module @com_google_protobuf//src/google/protobuf:protobuf_nowkt does not depend on a module exporting 'absl/strings/strip.h'
external/com_google_protobuf/src/google/protobuf/descriptor.cc:49:10: error: module @com_google_protobuf//src/google/protobuf:protobuf_nowkt does not depend on a module exporting 'absl/strings/substitute.h'
external/com_google_protobuf/src/google/protobuf/descriptor.cc:51:10: error: module @com_google_protobuf//src/google/protobuf:protobuf_nowkt does not depend on a module exporting 'absl/types/optional.h'
external/com_google_protobuf/src/google/protobuf/descriptor.cc:72:10: error: module @com_google_protobuf//src/google/protobuf:protobuf_nowkt does not depend on a module exporting 'google/protobuf/port_def.inc'
... and so on ...
```

Say there was a hypothetical tool `build_cleaner` that adds all the dependencies in a BUILD file where it sees an include for something but not the dependency. That could maybe fix it.

**Comment by hzeller** [comment]
Looks like this had been fixed in https://github.com/protocolbuffers/protobuf/commit/a7b0421c78412baa48704d727601a17ac0f451d1 and is
 resolved as of 26.0-rc1 (but not in 25.x yet).

Thanks @acozzette
